waku/v2/protocol/waku_peer_exchange/protocol.nim

```diff
@@ -147,44 +130,34 @@ proc getEnrsFromCache(wpx: WakuPeerExchange, numPeers: uint64): seq[enr.Record]
     return @[]
   for i in 0..<min(numPeers, wpx.enrCache.len().uint64()):
     let ri = rand(0..<wpx.enrCache.len())
+    # TODO: Note that duplicated peers can be returned here
     result.add(wpx.enrCache[ri])
 
 proc initProtocolHandler(wpx: WakuPeerExchange) =
   proc handler(conn: Connection, proto: string) {.async, gcsafe, closure.} =
-    let buff = await conn.readLp(MaxRpcSize.int)
+    var buffer: seq[byte]
+    try:
+      buffer = await conn.readLp(MaxRpcSize.int)
+    except CatchableError as exc:
+      waku_px_errors.inc(labelValues = [exc.msg])
+      return
 
-    let res = PeerExchangeRpc.decode(buff)
+    let res = PeerExchangeRpc.decode(buffer)
     if res.isErr():
       waku_px_errors.inc(labelValues = [decodeRpcFailure])
       return
 
     let rpc = res.get()
 
-    # handle peer exchange request
-    if rpc.request != PeerExchangeRequest():
+    # If we got a request (request field is not empty)
+    if rpc.request != default(PeerExchangeRequest):
```

Comment:
While I know this is in other places in the code, we should avoid doing it. This is a defect in this protocol's RPC codec and RPC types: 

- If the field is mandatory, return a "required field missing error"
- If the field is optional, the codec should return an `Option`; here, you should check if it is present (or not) and act in consequence.

Reply:
Yep, I agree. But since the rpc doesn't allow us to add an error message, this would involve changing the px proto, changing the spec, and perhaps bumping the px version? All of this beyond the PR.

Reply:
This is not correct. 

I am not asking to change the protocol. I am concerned about the implementation and the way this equality comparison uses the `default(...)` to detect if the field was missing:

```Nim
    if rpc.request != default(PeerExchangeRequest):
```

You can return the [this "required field missing" error](https://github.com/waku-org/nwaku/blob/master/waku/v2/protocol/waku_store/rpc_codec.nim#L63) from the `rpc_codec` the same way other protocols do, and this check can be removed from the protocol handler because if the field is not present, the message will be considered invalid.

This doesn't require any protocol modification. It is an implementation detail, and it should be in the scope of the PR given that this PR aims to improve the implementation.

Reply:
Should be fixed in 7cb54339fe20cf991a9816a62e6447b94ae13212

> and it should be in the scope of the PR

note thought that the initial goal of this PR was to close https://github.com/waku-org/nwaku/issues/1463 and some refactoring was done along the way, but the priority was to decouple discovery from connection.

Reply:
As a general note, PRs can and should have a limited scope. This PR had a clear focus listed in the description and addressed those items.

Not approving a PR generally means that there is a critical bug/regression/technical debt _introduced_ (not existing prior to PR) or a critical question which needs answering. Technical debt which existed before the PR can be pointed out, but to speed up PR process should not stop approval.